src/controller/CHIPDeviceController.cpp

```diff
@@ -841,6 +841,16 @@ CHIP_ERROR DeviceCommissioner::PairDevice(NodeId remoteDeviceId, RendezvousParam
     VerifyOrExit(mDeviceBeingPaired < kNumMaxActiveDevices, err = CHIP_ERROR_NO_MEMORY);
     device = &mActiveDevices[mDeviceBeingPaired];
 
+    // If the CSRNonce is passed in, using that else generating one to be used for commissioning.
+    if (params.HasCSRNonce())
+    {
+        device->SetCSRNonce(params.GetCSRNonce());
+    }
+    else
+    {
+        device->GenerateCSRNonce();
```

Comment:
We could get rid of this generate: instead the controller does the Rand() and the only method in the device is 'SetCSR'.

Reply:
The details like kOpCSRNonceLength on which the Generate method depend seem like they should be private to CHIPDevice and controller should not worry about it hence called the 2 methods separately.

Reply:
Generating it in the commissioner app should be good. We can make it a mandatory init parameter, instead of optional.

Reply:
Added a TODO for it as the current flow relies on the auto generated CSRNonce and if we abstract the generation into the controller to always provide one to the CHIPDevice, we also need to update the implementation.  Its a little out of the scope for this PR hence adding a TODO.

Reply:
I am not asking for commisioner (long term would be ideal), for now I only ask the CHIPDevice to only expose 'SetCSRNonce' and not have a 'GenerateCSRNonce' command. 

The code that is  currently in GenerateCSRNonce can be moved in here. Having a single 'Set'  seems prefferable as a smaller API surface.

Reply:
Just realized that although this file is CHIPDeviceController we are changing 'ChipDeviceCommisioner'.

So rephrasing: I am not asking for init param changes in this PR, but I am asking for the random to be taken out and be implemented in this file for now rather than in CHIPDevice.

Reply:
done